### DISCUSSION TITLE
Geometric stiffness (Jacobian) for Pressure BC

### DISCUSSION THREAD

**Discussion by matthiasneuner** [idaholab/moose#19566] [Q&A Modules: Solid mechanics] [answered]
Hi, it seems to me that for the `Pressure` BC in the `TensorMechanics` module the Jacobian wrt the displacement variables is missing, despite the fact that this BC acts on the deformed mesh.
I am simulating materials with very large deformations, and accordingly, this Jacobian is essential for me.
A quick&dirty implementation by implementing the derivative of the boundary `_normals` wrt the displacement variables in `computeQpJacobian()` in `Pressure` got me the desired behavior.

Now for cleaning up that workaround and pushing that to the MOOSE main repo, I was wondering if `Pressure` is actually the right place to implement the derivative of the `_normals` wrt the displacement vector, or if this should be performed at some higher level, since essentially all types of follower loads may use that for computing the exact Jacobian. 
If so, what would be the ideal place to compute that derivative?

**Comment by jiangwen84** [comment, marked as answer]
@matthiasneuner You are correct that the Jacobian in Pressure BC is not exact. `normal` and `area` changes with respect to displacements are not considered. Have you checked your hand-coded Jacobian against finite difference approximation? Are they exact?  

Let me tag @jasondhales since he recently worked on it. He might give you some suggestion on where the derivatives calculation should go.

**Reply by matthiasneuner** [reply]
My Jacobian is exact yes. You can find my implementation in my app (https://github.com/matthiasneuner/chamois/blob/master/src/bcs/FiniteStrainPressure.C), making use of an auxiliary material (https://github.com/matthiasneuner/chamois/blob/master/src/materials/ComputeDeformedBoundaryNormalVector.C). 

Since I use Fastor for tensor contractions instead of MOOSE tensors, this implementation is not yetcompatible with the main repo. Maybe I can rewrite that using MOOSE tensors during the holidays, and create a PR.

**Reply by jiangwen84** [reply]
I took a quick look at your implementation of pressure BC. You set `use_displaced_mesh = false` which means the pressure term is integrated over the undeformed configuration. Shouldn't we account for the area change due to deformation?

**Reply by matthiasneuner** [reply]
You are right; This is actually happening; Nanson's formula is applied in ComputeDeformedBoundaryNormalVector (https://github.com/matthiasneuner/chamois/blob/6066b7dfc4d9456dd971f2a91b470899fe98fe78/src/materials/ComputeDeformedBoundaryNormalVector.C#L79), but I just recognized that the name is misleading since the vector is not a normalized vector anymore. I will improve that.

**Reply by hugary1995** [reply]
I think this ought to be handled in "Assembly". For example, now in `ADIntegratedBC` we have
```cpp
_normals(_assembly.adNormals()),
```
and in `IntegratedBC` we have
```cpp
_normals(_assembly.normals()),
```

I think we should add a method to compute the derivative of normal w.r.t to displacements in assembly, and manually request it when needed (for example in the case of pressure BC):
```cpp
_d_normals_d_displacements(_assembly.dNormalsDDisplacements()),
```

In the case of AD, these derivatives should come from AD libMesh. 

@lindsayad is this a use case of AD libMesh?

**Reply by lindsayad** [reply]
You mean the AD libmesh that doesn't exist? 😄 Yes as @hugary1995 pointed out we do currently have "manual" code in `Assembly` that reproduces mapping calculations from libMesh with displacement information included in order to create an `adNormals()` that has the correct derivatives with respect to displacements. Someday when we do have AD libMesh, then we can remove that near-duplicate code in `Assembly`

But as far as creating derivatives with respect to displacements for our manual Jacobian objects, I will leave that to you all 😄

**Comment by jasondhales** [comment]
My take is that the derivative code ought to go into `Pressure.C`.  Getting the derivative of the `normal` would be a good contribution.
